addons/test_mail/tests/test_mail_gateway.py

```diff
@@ -801,6 +801,73 @@ def test_message_route_alias_multiple_new(self):
         new_simple = self.env['mail.test.gateway'].search([('name', '=', 'Test Subject')])
         self.assertEqual(len(new_simple), 1, 'message_process: a new mail.test should have been created')
 
+    @mute_logger('odoo.addons.mail.models.mail_thread', 'odoo.models')
+    def test_mail_catchall_domain_allowed_validation(self):
+        """ Check the validation of `mail.catchall.domain.allowed` system parameter"""
+        for value in (',', ',,'):
+            with self.assertRaises(exceptions.ValidationError):
+                self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', value)
+
+        for value in ('', 'hello.com', 'hello.com,,', 'hello.com,bonjour.com', 'hello.com, bonjour.com'):
+            self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', value)
+            value = ",".join([domain.strip() for domain in value.split(',') if domain]) if value else False
+            self.assertEqual(self.env['ir.config_parameter'].get_param('mail.catchall.domain.allowed'), value)
+
+    @mute_logger('odoo.addons.mail.models.mail_thread', 'odoo.models')
+    def test_message_route_alias_with_allowed_domains(self):
+        """ Incoming email: check that if domains are set in the
+        optional system parameter `mail.catchall.domain.allowed`,
+        only incoming emails from these domains will generate records."""
+
+        MailTestGatewayModel = self.env['mail.test.gateway']
+        MailTestContainerModel = self.env['mail.test.container']
+
+        allowed_domain = 'hello.com'
+        not_allowed_domain = 'bonjour.com'
+
+        # test@.. will cause the creation of new mail.test
+        new_alias_2 = self.env['mail.alias'].create({
+            'alias_name': 'test',
+            'alias_user_id': False,
+            'alias_model_id': self.env['ir.model']._get('mail.test.container').id,
+            'alias_contact': 'everyone',
+        })
+
+        for subject, val1, val2, domain, sys_param in [
+            # Test with 'mail.catchall.domain.allowed' not set in system parameters
+            # and with a domain not allowed
+            ('Test Subject 1', 1, 1, not_allowed_domain, ""),
+            # Test with 'mail.catchall.domain.allowed' set in system parameters
+            # and with a domain not allowed
+            ('Test Subject 2', 1, 0, not_allowed_domain, allowed_domain),
+            # Test with 'mail.catchall.domain.allowed' set in system parameters
+            # and with a domain allowed
+            ('Test Subject 3', 1, 1, allowed_domain, allowed_domain),
+        ]:
+            self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', sys_param)
```

Comment:
Note that you could use "with self.subTest(...):", so that if one test crashes, the other tests are launched. It allows to test all use cases without having to fix them one by one, usefull in refactorings :)  you may grep existing code for some examples (you may give parameters to subtest, like ``with self.subTest(alias2_domain=alias2_domain, ...):``, and those are notably displayed in the console, allowing to know what are the parameters of the subtest)

Reply:
Sorry, when I refactored the loop with your previous comment, I forgot to add the `with self.subTest`, my bad, thanks :pray: